composer/callbacks/export_for_inference.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Callback to export model for inference."""
+
+from __future__ import annotations
+
+import logging
+from typing import Callable, Optional, Sequence, Union
+
+import torch.nn as nn
+
+from composer.core import State
+from composer.core.callback import Callback
+from composer.loggers import Logger
+from composer.utils.inference import ExportFormat, export_for_inference
+from composer.utils.object_store import ObjectStore
+
+log = logging.getLogger(__name__)
+
+__all__ = ['ExportForInference']
+
+
+class ExportForInference(Callback):  # noqa: D101
+    """Callback to export model for inference.
+
+    Args:
+        save_format (Union[str, ExportFormat]):  Format to export to. Either ``"torchscript"`` or ``"onnx"``.
+        out_filename (str): The path for storing the exported model. It can be a path to a file on the local disk,
+            a URL, or if ``save_object_store`` is set, the object name
+            in a cloud bucket. For example, ``my_run/exported_model``.
+        save_object_store (ObjectStore, optional): If the ``out_filename`` is in an object name in a cloud bucket
+            (i.e. AWS S3 or Google Cloud Storage), an instance of
+            :class:`~.ObjectStore` which will be used
+            to store the exported model. Set this to ``None`` if ``out_filename`` is a local filepath.
+            (default: ``None``)
+        transforms (Union[Callable, Sequence[Callable]], optional): transformations (usually optimizations) that should
+            be applied to the model. Each should be a callable that takes a model and returns a modified model.
+    """
+
+    def __init__(
+        self,
+        save_format: Union[str, ExportFormat],
+        out_filename: str,
+        save_object_store: Optional[ObjectStore] = None,
+        transforms: Optional[Union[Callable[[nn.Module], nn.Module], Sequence[Callable[[nn.Module],
```

Comment:
Could we simplify the typing here with an alias? e.g., up top to 
```
Transform = Callable[[nn.Module],nn.Module]
```

then

```
transforms: Union[Transform, Sequence[Transform]]
```

Reply:
We can definitely do that. However, this still won't fix the error I'm currently seeing when running test_callback_hparams_is_constructable
`TypeError: A union of multiple classes without a primitive or Enum are not supported`

Reply:
Ah must be a `yahp` limitation. Can you try `Union[Callable[], Callable[], None]`?

Reply:
Unfortunately, yahp defines [primitive types](https://github.com/mosaicml/yahp/blob/bbe1a826e7c492722199e303bc1868d47664a4dd/yahp/utils/type_helpers.py#L25) as bool, float, ints, or strings. So None does not work. If there's an ignore flag we can use to get yahp to skip this, that could work.

Reply:
Resolved by using Optional[Sequence[Transform]]